File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const n = maps.length;
    const m = maps[0].length;
    let visited = Array.from({ length: n }, () => Array(m).fill(false));
    
    const move = [[1, 0], [0, 1], [-1, 0], [0, -1]];
    let answer = 0;
    //queue에 저장할 때, 현재 위치와 현재 이동거리를 함께 저장
    let queue = [];
    queue.push([0, 0, 1]);
    visited[0][0] = true;
    
    while(queue.length > 0) {
        let [currRow, currCol, currDist] = queue.shift();
        
        if(currRow === n - 1 && currCol === m - 1) {
            return currDist;
        }
        
        for(let i = 0; i < move.length; i++) {
            let nextRow = currRow + move[i][0];
            let nextCol = currCol + move[i][1];
            
            if(nextRow >= 0 && nextRow < n && nextCol >= 0 && nextCol < m){
                if(maps[nextRow][nextCol] === 1 && !visited[nextRow][nextCol]){
                    visited[nextRow][nextCol] = true;
                    queue.push([nextRow, nextCol, currDist + 1])
                }
            }
        }
    }
    //queue를 다 돌았는데도 최단 거리를 구하지 못했다면 도달하지 못함
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
import collections

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    visited = [[False] * m for _ in range(n)]
    move = [(0, 1), (1,0), (-1, 0), (0, -1)]
    
    queue = collections.deque([(0, 0, 1)])
    visited[0][0] = True
    
    while queue:
        curr_row, curr_col, curr_dist = queue.popleft()
        
        if curr_row == n - 1 and curr_col == m - 1:
            return curr_dist
        
        for dr, dc in move:
            next_row, next_col = curr_row + dr, curr_col + dc
        
            if 0 <= next_row < n and 0 <= next_col < m:
                if maps[next_row][next_col] == 1 and not visited[next_row][next_col]:
                    visited[next_row][next_col] = True
                    queue.append((next_row, next_col, curr_dist + 1))
    
    return -1